### PR TITLE
logging: File system backend

### PR DESCRIPTION
**Pull request by nvlsianpu**
This PR introduces a new LOG backend based on VFS API. It is able to store Loggs in multiple files in pointed log directory

* cold start supported
If FS volume is not mounted from the beginning the backend will drop LOG. Backend will start logging normally once FS volume will be mounted (so FS can be directly mounted by the application either). Obviously the backend works well especially if FS volume is automounted within zephyr-rtos startup.
* auto recovery after unexpected lost of the current log file in runtime.
If currently used log file is lost, the backend attempt to create another one.

On the first possible write attempt backend creates a new file with continuous numeration. When max size of file is reached. 
system creates another one. There is a limit of log files used by backend too. There is a possibility to overwrite
old files or drop new ones.

smp_svr sample was modifed so it has enabled FS commonest. This project could be used to test new backend.
The easiest way to download the log file is using this command:
mcumgr --conntype serial --connstring='dev=/dev/ttyACM0,baud=115200' fs download /lfs1/log.0000 ~/log.txt.

Wort to note that a Log file can be also read using FS shell.

This PR supersedes https://github.com/zephyrproject-rtos/zephyr/pull/27486 as I took over the task and reworked base code much so want completely new review process.

### CODE REVIEW

**Comment by lukasbrchl** [comment]
Hello, this is a great feature and we look forward to using it. However, do you plan to add functionality to list files in a directory? Because right now, we wouldn't know which file to download.

**Comment by nvlsianpu** [comment]
> Hello, this is a great feature and we look forward to using it. However, do you plan to add functionality to list files in a directory? Because right now, we wouldn't know which file to download.

mcumgr miss that feature, however you can use FS shell `fs list` command. I don't have plan to work on mcumgr right now.

**Comment by koffes** [comment]
Thanks for the contribution, we will try this as well. Has it been tested with both FAT, exFAT and LittleFS?

File: tests/subsys/logging/log_backend_fs/src/log_fs_test.c [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+ */
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <zephyr.h>
+#include <ztest.h>
+#include <fs/fs.h>
+
+#define DT_DRV_COMPAT zephyr_fstab_littlefs
+#define TEST_AUTOMONT DT_PROP(DT_DRV_INST(0), automount)
```
Comment (nordic-krch) [inline tests/subsys/logging/log_backend_fs/src/log_fs_test.c:20]: TEST_AUTOMOUNT

**Review by nordic-krch — COMMENTED** [review]
From logging perspective it looks ok. One thing to consider in the future: you could store log messages in the future, especially after logging overhaul where log message will be self contained (will contains transient strings, too). It should take less space and be faster.

**Comment by nvlsianpu** [comment]
> Thanks for the contribution, we will try this as well. Has it been tested with both FAT, exFAT and LittleFS?

@koffes Was tested using LittleFS. Will work with any FS supported by VFS API.

**Review by nordic-krch — APPROVED** [review]
Reviewed only logging interface.

**Comment by nvlsianpu** [comment]
@aunsbjerg would be great if you can look at this.

**Review by nashif — APPROVED** [review]

**Pull request merged**